app/main/views/users.py

```diff
@@ -271,24 +282,45 @@ def export_users_for_framework(framework_slug):
 
     suppliers_with_a_complete_service = frozenset(framework.get_supplier_ids_for_completed_service())
 
-    supplier_id_published_service_count = dict(db.session.query(
-        Service.supplier_id,
-        func.count(Service.id)
-    ).filter(
-        Service.status == 'published',
-        Service.framework_id == framework.id
-    ).group_by(
-        Service.supplier_id
-    ).all())
+    lots = db.session.query(FrameworkLot.lot_id, Lot.slug).join(Lot, FrameworkLot.lot_id == Lot.id).filter(
+        FrameworkLot.framework_id == framework.id
+    ).distinct().all()
+    published_service_count_by_supplier_and_lot = {}
+
+    for lot_id, slug in lots:
+        published_service_count_by_supplier_and_lot[
+            "published_services_count_on_{}_lot".format(slug)
+        ] = dict(db.session.query(
+            Service.supplier_id,
+            func.count(Service.id)
+        ).filter(
+            Service.status == 'published',
+            Service.lot_id == lot_id
+        ).group_by(
+            Service.supplier_id
+        ).group_by(
+            Service.lot_id
+        ).all())
+
+    supplier_id_published_service_count = {}
+    for k, v in published_service_count_by_supplier_and_lot.items():
+        for supplier_id, service_count in v.items():
+            if supplier_id in supplier_id_published_service_count:
+                supplier_id_published_service_count[supplier_id] += service_count
+            else:
+                supplier_id_published_service_count[supplier_id] = service_count
+
```

Comment:
You're reinventing sql in slightly twisty-turny python. This is all doable in a single SQL statement.

Reply:
Whole thing more or less doable with two lines added:

```
db.session.query(
    Service.supplier_id,
    func.count(Service.id)
 ).filter(
    Service.status == 'published',
    Service.framework_id == framework.id
).group_by(
    Service.supplier_id
).all()
```

becomes

```
db.session.query(
    Service.supplier_id,
    Service.lot_id,
    func.count(Service.id)
 ).filter(
    Service.status == 'published',
    Service.framework_id == framework.id
).group_by(
    Service.supplier_id,
    Service.lot_id,
).all()
```

Reply:
Though there's actually an even neater way of doing it - let me think...